control/noise/handshake.go

```diff
@@ -0,0 +1,430 @@
+	}
+	if err := s.MixDH(controlKey, machineKey); err != nil {
+		return nil, fmt.Errorf("computing ss: %w", err)
+	}
+	if err := s.DecryptAndHash(nil, init.Tag()); err != nil {
+		return nil, fmt.Errorf("decrypting initiation tag: %w", err)
+	}
+
+	// <- e, ee, se
+	resp := mkResponseMessage()
+	controlEphemeral := key.NewPrivate()
+	controlEphemeralPub := controlEphemeral.Public()
+	copy(resp.EphemeralPub(), controlEphemeralPub[:])
+	s.MixHash(controlEphemeralPub[:])
+	if err := s.MixDH(controlEphemeral, machineEphemeralPub); err != nil {
+		return nil, fmt.Errorf("computing ee: %w", err)
+	}
+	if err := s.MixDH(controlEphemeral, machineKey); err != nil {
+		return nil, fmt.Errorf("computing se: %w", err)
+	}
+	s.EncryptAndHash(resp.Tag(), nil) // empty message payload
+
+	c1, c2, err := s.Split()
+	if err != nil {
+		return nil, fmt.Errorf("finalizing handshake: %w", err)
+	}
+
+	if _, err := conn.Write(resp[:]); err != nil {
+		return nil, err
+	}
+
+	return &Conn{
+		conn:          conn,
+		version:       protocolVersion,
+		peer:          machineKey,
+		handshakeHash: s.h,
+		tx: txState{
+			cipher: c2,
+		},
+		rx: rxState{
+			cipher: c1,
+		},
+	}, nil
+}
+
+// symmetricState is the SymmetricState object from the Noise protocol
+// spec. It contains all the symmetric cipher state of an in-flight
+// handshake. Field names match the variable names in the spec.
+type symmetricState struct {
+	h  [blake2s.Size]byte
+	ck [blake2s.Size]byte
+
+	k [chp.KeySize]byte
+	n uint64
+
+	mixer hash.Hash // for updating h
+}
+
+// Initialize sets s to the initial handshake state, prior to
+// processing any Noise messages.
+func (s *symmetricState) Initialize() {
+	if s.mixer != nil {
+		panic("symmetricState cannot be reused")
+	}
+	s.h = blake2s.Sum256([]byte(protocolName))
+	s.ck = s.h
+	s.k = [chp.KeySize]byte{}
+	s.n = invalidNonce
+	s.mixer = newBLAKE2s()
+}
+
+// MixHash updates s.h to be BLAKE2s(s.h || data), where || is
+// concatenation.
+func (s *symmetricState) MixHash(data []byte) {
+	s.mixer.Reset()
+	s.mixer.Write(s.h[:])
+	s.mixer.Write(data)
+	s.mixer.Sum(s.h[:0]) // TODO: check this actually updates s.h correctly...
+}
+
+// MixDH updates s.ck and s.k with the result of X25519(priv, pub).
+//
+// MixDH corresponds to MixKey(X25519(...))) in the spec. Implementing
+// it as a single function allows for strongly-typed arguments that
+// reduce the risk of error in the caller (e.g. invoking X25519 with
+// two private keys, or two public keys), and thus producing the wrong
+// calculation.
+func (s *symmetricState) MixDH(priv key.Private, pub key.Public) error {
+	// TODO(danderson): check that this operation is correct. The docs
+	// for X25519 say that the 2nd arg must be either Basepoint or the
+	// output of another X25519 call.
+	//
+	// I think this is correct, because pub is the result of a
+	// ScalarBaseMult on the private key, and our private key
+	// generation code clamps keys to avoid low order points. I
```

Comment:
`ScalarBaseMult` is identical to `X25519(..., Basepoint)`, and they both clamp. You don't need to do clamping in your own code with either.

The code is indeed correct.